pkg/batchmapper/service_test.go

```diff
@@ -0,0 +1,249 @@
+package batchmapper
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"reflect"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	batchmappb "github.com/numaproj/numaflow-go/pkg/apis/proto/batchmap/v1"
+)
+
+type BatchMapStreamFnServerTest struct {
+	ctx      context.Context
+	outputCh chan *batchmappb.BatchMapResponse
+	inputCh  chan *batchmappb.BatchMapRequest
+	grpc.ServerStream
+}
+
+func NewBatchBatchMapStreamFnServerTest(
+	ctx context.Context,
+	inputCh chan *batchmappb.BatchMapRequest,
+	outputCh chan *batchmappb.BatchMapResponse,
+) *BatchMapStreamFnServerTest {
+	return &BatchMapStreamFnServerTest{
+		ctx:      ctx,
+		inputCh:  inputCh,
+		outputCh: outputCh,
+	}
+}
+
+func (u *BatchMapStreamFnServerTest) Recv() (*batchmappb.BatchMapRequest, error) {
+	val, ok := <-u.inputCh
+	if !ok {
+		return val, io.EOF
+	}
+	return val, nil
+}
+
+func (u *BatchMapStreamFnServerTest) Send(d *batchmappb.BatchMapResponse) error {
+	u.outputCh <- d
+	return nil
+}
+
+func (u *BatchMapStreamFnServerTest) Context() context.Context {
+	return u.ctx
+}
+
+type BatchMapFnServerErrTest struct {
+	ctx      context.Context
+	inputCh  chan *batchmappb.BatchMapRequest
+	outputCh chan *batchmappb.BatchMapResponse
+	grpc.ServerStream
+}
+
+func NewBatchMapFnServerErrTest(
+	ctx context.Context,
+	inputCh chan *batchmappb.BatchMapRequest,
+	outputCh chan *batchmappb.BatchMapResponse,
+
+) *BatchMapFnServerErrTest {
+	return &BatchMapFnServerErrTest{
+		ctx:      ctx,
+		inputCh:  inputCh,
+		outputCh: outputCh,
+	}
+}
+
+func (u *BatchMapFnServerErrTest) Recv() (*batchmappb.BatchMapRequest, error) {
+	val, ok := <-u.inputCh
+	if !ok {
+		return val, io.EOF
+	}
+	return val, nil
+}
+
+func (u *BatchMapFnServerErrTest) Send(_ *batchmappb.BatchMapResponse) error {
+	return fmt.Errorf("send error")
+}
+
+func (u *BatchMapFnServerErrTest) Context() context.Context {
+	return u.ctx
+}
+
+func TestService_BatchMapFn(t *testing.T) {
+	tests := []struct {
+		name        string
+		handler     BatchMapper
+		input       []*batchmappb.BatchMapRequest
+		expected    []*batchmappb.BatchMapResponse
+		expectedErr bool
+	}{
```

Comment:
Can we add a test to cover the error/panic case inside the user code?

Reply:
In the current setup, this is difficult to test via UT directly. I tried to add the test there but the assert.Panics (which internally uses recover) was not able to catch it as the panic is invoked from within another goroutine. Would need to research more on how can we cover something like this. In python i'm able to mock the panic and test which wasn't working out here.

I tried few manual scenarios to induce such panics which were working fine.